lib: location: Add new Location library (#5963)
nrfconnect/sdk-nrf: doc/nrf/libraries/modem/location.rst

.. _lib_location:

Location
########

.. contents::
   :local:
   :depth: 2

The location library provides functionality for retrieving the location of a device using different positioning methods such as:

 * GNSS satellite positioning including Assisted GPS (A-GPS) and Predicted GPS (P-GPS) data.
 * Cellular positioning.
 * Wi-Fi positioning.

Overview
********

> This library provides an API for clients to request the location of a device.

Author: Client is the one who uses the library and that could be an application or another library. Should I use application instead although that's a bit inaccurate.
`REST client` is obviously different and the word client is part of the name of that library so cannot do anything. I comes from the references and not that we would call that library with the word client.

> This library provides an API for clients to request the location of a device.

Author: @simensrostad Can you comment on this one? Should I use "application" for the SW component that uses Location library?

> This library provides an API for clients to request the location of a device.

Reviewer: I think that would clear things up a bit.

.. _lib_location:

Location
########

.. contents::
   :local:
   :depth: 2

Location library provides functionality for retrieving device's location utilizing different positioning methods
such as GNSS satellite positioning including Assisted GPS (A-GPS) and predicted GPS (P-GPS) data, cellular positioning or WiFi positioning.

Overview
********

Location library provides a convenient API for clients to request devices location.
The client can determine the preference order of the positioning methods to be used along with some other configuration information.
Once a method fails to provide location, Location library will perform fallback to the next preferred method.

> Supported positioning methods are GNSS, cellular and WiFi positioning.

Author: Yep, deleting.

> Both cellular and WiFi positioning take advantage of the base stations seen with corresponding technology, and then utilizing web services for retrieving the location.

Author: I did use Pekka's suggestion here.

GNSS positioning uses satellites to compute the location of the device. Location library can utilize assistance data (A-GPS and P-GPS) to find the satellites faster.

Implementation
==============

> Location library has a compact API and then location core that handles location method independent part of the functionality such as fallback to the next preferred method and timeouts.

Reviewer 2: ```suggestion
Location library has a compact API and a core that handles location method independent part of the functionality such as fallback to the next preferred method and timeouts.
```

Each location method has its own implementation for the location retrieval:

* GNSS positioning

>    * :ref:`gnss_interface` for getting satellite information

Author: Changing this to:
gnss_interface for getting the location.

   * A-GPS/P-GPS are managed either with
      * :ref:`lib_nrf_cloud_agps` and :ref:`lib_nrf_cloud_pgps`, or
      * the application utilizes some other source for the data and uses :c:func:`location_agps_data_process` to pass the data to Location library

>    * A-GPS/P-GPS data format must be as received from :ref:`lib_nrf_cloud_agps`

Reviewer 2: This line provides details to the line above, so it could be indented two times to right. Also, the external assistance can only be A-GPS, not P-GPS because of the nRF P-GPS library limitations.

Author: Pekka moved the bullets to the same level so I'm not doing anything for this.

   * A-GPS/P-GPS data transport for :ref:`lib_nrf_cloud_agps` and :ref:`lib_nrf_cloud_pgps` can be configured to be either MQTT (:kconfig:`CONFIG_NRF_CLOUD_MQTT`) or REST (:kconfig:`CONFIG_NRF_CLOUD_REST`)
* Cellular positioning
   * :ref:`lte_lc_readme` for getting visible cellular base stations
   * :ref:`lib_multicell_location` for sending cell information to the selected location service and getting calculated location back to the device
      * Service is selected in :c:struct:`loc_method_config` when requesting location
      * Data transport for the service is REST
      * Available services are `nRF Cloud Location Services`_, `HERE Positioning`_, `Skyhook Precision Location`_ and `Polte Location API`_
* WiFi positioning
   * Zephyr's Network Management API :ref:`zephyr:net_mgmt_interface` for getting visible WiFi access points

>    * Sending access point information to the selected location service and getting calculated location back to the device

Reviewer 2: ```suggestion
   * Send access point information to the selected location service and get calculated location back to the device
```

      * Location library has implementation for WiFi location services
      * Service is selected in :c:struct:`loc_method_config` when requesting location
      * Data transport for the service is REST
      * Available services are `nRF Cloud Location Services`_, `HERE Positioning`_ and `Skyhook Precision Location`_

Supported features
==================

> TODO: Not really sure what to put into this mandatory section given Implementation and Configuration sections has a lot of information and if something is missing, that can be added.

Author: @peknis You could comment this one too. I can repeat some stuff from Overview and Implementation sections but is there something different that I should write here?

Author: Removing Supported features section.

.. note::
   Use this section to describe the features supported by the library.

Requirements
************

nRF Cloud certificates
======================

If you use nRF Cloud for any location data, you must have the certificate provisioned.
See `Updating the nRF Connect for Cloud certificate`_ for more information.

Location service accounts
=========================

> To use the location services that provide A-GPS/P-GPS, cellular positioning and WiFi positioning data, see the respective documentation for account setup and for getting the required credentials for authentication:

Reviewer 2: ```suggestion
To use the location services that provide A-GPS/P-GPS, cellular positioning or WiFi positioning data, see the respective documentation for account setup and for getting the required credentials for authentication:
```

> To use the location services that provide A-GPS/P-GPS, cellular positioning and WiFi positioning data, see the respective documentation for account setup and for getting the required credentials for authentication:

Reviewer 3: ```suggestion
To use the location services that provide A-GPS or P-GPS, cellular and Wi-Fi positioning data, see the respective documentation for setting up your account and getting the required credentials for authentication:
```

* `nRF Cloud Location Services`_
* `HERE Positioning`_
* `Skyhook Precision Location`_
* `Polte Location API`_

> The required credentials for the location services are configurable using Kconfig options.

Reviewer 3: ```suggestion
You can configure the required credentials for the location services using Kconfig options.
```

> WiFi chip
> =========

Reviewer 3: ```suggestion
Wi-Fi chip
==========
```

> WiFi is not supported by the HW of the supported DKs. External WiFi chips can be used such as ESP32, which you can connect to the DKs.

Reviewer 3: ```suggestion
None of the supported DKs have a Wi-Fi chip. You can use external Wi-Fi chips, such as ESP32, and connect them to the DKs.
```

> TODO: any links we should add?

Reviewer 4: Sample and a device tree overlay is having some instructions: https://github.com/nrfconnect/sdk-nrf/blob/7d11429c4d5eab0a88de34af1b6262678cbfede3/samples/nrf9160/location/esp_8266_nrf9160ns.overlay#L7

> TODO: any links we should add?

Author: ESP32 probably came while browsing through Internet. Fixed that.
I'm now referring to the sample and its overlay for integration information. Other than that, I guess it's out of the scope of this document.

Library files
*************

.. |library path| replace:: :file:`lib/location`

This library is found under |library path| in the |NCS| folder structure.

Configuration
*************

Configure the following Kconfig option to enable this library:

* :kconfig:`CONFIG_LOCATION` - Enables the Location library.

Configure the following options to enable location methods of your choice:

* :kconfig:`CONFIG_LOCATION_METHOD_GNSS` - Enables GNSS location method.
* :kconfig:`CONFIG_LOCATION_METHOD_CELLULAR` - Enables cellular location method.

> * :kconfig:`CONFIG_LOCATION_METHOD_WIFI` - Enables WiFi location method.
> * :kconfig:`CONFIG_WIFI` - Enable WiFi for Zephyr.

Reviewer 3: ```suggestion
* :kconfig:`CONFIG_LOCATION_METHOD_WIFI` - Enables Wi-Fi location method.
* :kconfig:`CONFIG_WIFI` - Enable Wi-Fi for Zephyr.
```

The following options control the use of GNSS assistance data:

* :kconfig:`CONFIG_LOCATION_METHOD_GNSS_AGPS_EXTERNAL` - Enables A-GPS data retrieval from an external source which the application implements separately.
* :kconfig:`CONFIG_NRF_CLOUD_AGPS` - Enables A-GPS data retrieval from `nRF Cloud`_.
* :kconfig:`CONFIG_NRF_CLOUD_PGPS` - Enables P-GPS data retrieval from `nRF Cloud`_.

> The following options control the transport used with `nRF Cloud`_:

Reviewer 3: ```suggestion
The following options control the transport method used with `nRF Cloud`_:
```

* :kconfig:`CONFIG_NRF_CLOUD_REST` - Uses REST APIs to communicate with `nRF Cloud`_.
* :kconfig:`CONFIG_NRF_CLOUD_MQTT` - Uses MQTT transport to communicate with `nRF Cloud`_.

> * :kconfig:`CONFIG_REST_CLIENT` - Enable :ref:`lib_rest_client` library.

Reviewer 2: This is not currently used for GNSS method. Should we clarify here what is this good for? Similarly, can MQTT be used for all methods (cellular, wifi?)

Author: Implementation section comments about where REST and MQTT can be used so add a comment there if something is not true.

Author: @hiltunent Please close this or add further questions.

> Both cellular and WiFi location services are selected utilizing runtime configuration but the available services must be configured first.
> For cellular location services, use at least one of the following sets of options and configure corresponding authentication parameters (more details and configuration options can be found from :ref:`lib_multicell_location`):

Reviewer 3: ```suggestion
Both cellular and Wi-Fi location services are selected using the runtime configuration but the available services must be configured first.
For cellular location services, use at least one of the following sets of options and configure corresponding authentication parameters (for more details and configuration options, see :ref:`lib_multicell_location`):
```

* :kconfig:`CONFIG_MULTICELL_LOCATION_SERVICE_NRF_CLOUD`
* :kconfig:`CONFIG_MULTICELL_LOCATION_SERVICE_HERE` and :kconfig:`CONFIG_MULTICELL_LOCATION_HERE_API_KEY` (see below other authentication options)
* :kconfig:`CONFIG_MULTICELL_LOCATION_SERVICE_SKYHOOK` and :kconfig:`CONFIG_MULTICELL_LOCATION_SKYHOOK_API_KEY`
* :kconfig:`CONFIG_MULTICELL_LOCATION_SERVICE_POLTE` and :kconfig:`CONFIG_MULTICELL_LOCATION_POLTE_CUSTOMER_ID` and :kconfig:`CONFIG_MULTICELL_LOCATION_POLTE_API_TOKEN`

> For WiFi location services, use at least one of the following sets of options and configure corresponding authentication parameters:

Reviewer 3: ```suggestion
For Wi-Fi location services, use at least one of the following sets of options and configure the corresponding authentication parameters:
```

* :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_NRF_CLOUD`
* :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_HERE` and :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_HERE_API_KEY`
* :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_SKYHOOK` and :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_SKYHOOK_API_KEY`

> Following WiFi service related options can usually have default values:

Reviewer 3: ```suggestion
The following options are related to the Wi-Fi service and can usually have the default values:
```

* :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_HERE_HOSTNAME`
* :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_HERE_TLS_SEC_TAG`
* :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_SKYHOOK_HOSTNAME`
* :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_SKYHOOK_TLS_SEC_TAG`

Usage
*****

> Using Location library is rather easy. First you need to initialize the library with :c:func:`location_init`.

Reviewer 3: ```suggestion
To use the Location library, first initialize the library with the :c:func:`location_init` function.
```

Secondly, you need to set configuration (:c:struct:`loc_config`) including location method configurations (:c:struct:`loc_method_config`).

> This is achieved easily by setting first default values by using :c:func:`loc_config_defaults_set` and c:func:`loc_config_method_defaults_set`,

Reviewer 2: ```suggestion
This is easily achieved by setting first default values by using :c:func:`loc_config_defaults_set` and c:func:`loc_config_method_defaults_set`,
```

Author: Not relevant after Pekka's modification which I'll apply

> Secondly, you need to set configuration (:c:struct:`loc_config`) including location method configurations (:c:struct:`loc_method_config`).
> This is achieved easily by setting first default values by using :c:func:`loc_config_defaults_set` and c:func:`loc_config_method_defaults_set`,
> and then setting any required non-default values to the structures.

Reviewer 3: ```suggestion
Then, set the configuration (:c:struct:`loc_config`) including the location method (:c:struct:`loc_method_config`).
Set first the default values by using :c:func:`loc_config_defaults_set` and c:func:`loc_config_method_defaults_set`,
and then any required non-default values to the structures.
```

> Once the configuration is set up properly, you just need to call c:func:`location_request` with the configuration.

Reviewer 3: ```suggestion
Once you have set up the configuration properly, call the c:func:`location_request` function with the configuration.
```

Samples using the library
*************************

The following |NCS| samples use this library:

> * :ref:`location_sample`

Author: @peknis I guess I need your help here. This reference causes document build to fail:
`/home/runner/work/sdk-nrf/sdk-nrf/ncs/nrf/doc/_build/nrf/src/libraries/modem/location.rst:157:Failed to create a cross reference. A title or caption not found: location_sample`

README.rst of Location sample has this line in the beginning:
`.. _location_sample:`

Is the problem that these are both new files and the build is not able to find it because of that? I hope this is a typo that I still cannot see...

Author: This is fixed now by moving the |sample path| to correct place. It broke the link. Thanks Pekka!

* :ref:`modem_shell_application`

Limitations
***********

> * Location library can only have one client registered at a time. If there is already a client handler registered, another initialization will override the existing handler.
> * Cellular neighbor information used for cellular positioning is more accurate on Modem FW (MFW) 1.3.0 compared to earlier MFW releases which do not have API for scanning the neighboring cells.
>   For older than MFW 1.3.0 releases, only serving cell information is provided and that information may be from hours or days ago, or even older, depending on modem sleep states.

Reviewer 3: ```suggestion
* The Location library can only have one client registered at a time. If there is already a client handler registered, another initialization will override the existing handler.
* Cellular neighbor information used for cellular positioning is more accurate on modem firmware (MFW) 1.3.0 compared to earlier MFW releases that do not have an API for scanning the neighboring cells.
  For MFW releases older than 1.3.0, only serving cell information is provided and it can be hours or days old, or even older, depending on the modem sleep states.
```

Dependencies
************

This library uses the following |NCS| libraries:

* :ref:`nrf_modem_lib_readme`
* :ref:`lte_lc_readme`
* :ref:`lib_multicell_location`
* :ref:`lib_rest_client`
* :ref:`lib_nrf_cloud`
* :ref:`lib_nrf_cloud_agps`
* :ref:`lib_nrf_cloud_pgps`
* :ref:`lib_nrf_cloud_rest`

> 

Reviewer 2: modem_info and date_time libraries missing

Author: Added all these.

Author: Actually cannot find proper target to link for the cjson_lib because it's in modules and not NRF or Zephyr.

> It uses the following `sdk-nrfxlib`_ libraries:

Reviewer 3: ```suggestion
It uses the following `sdk-nrfxlib`_ library:
```

* :kconfig:`CONFIG_WIFI` - Enable WiFi for Zephyr.

The following options control the use of GNSS assistance data:

* :kconfig:`CONFIG_LOCATION_METHOD_GNSS_AGPS_EXTERNAL` - Enables A-GPS data retrieval from an external source which the application implements separately.
* :kconfig:`CONFIG_NRF_CLOUD_AGPS` - Enables A-GPS data retrieval from `nRF Cloud`_.
* :kconfig:`CONFIG_NRF_CLOUD_PGPS` - Enables P-GPS data retrieval from `nRF Cloud`_.

The following options control the transport used with `nRF Cloud`_:

* :kconfig:`CONFIG_NRF_CLOUD_REST` - Uses REST APIs to communicate with `nRF Cloud`_.
* :kconfig:`CONFIG_NRF_CLOUD_MQTT` - Uses MQTT transport to communicate with `nRF Cloud`_.
* :kconfig:`CONFIG_REST_CLIENT` - Enable :ref:`lib_rest_client` library.

Both cellular and WiFi location services are selected utilizing runtime configuration but the available services must be configured first.
For cellular location services, use at least one of the following sets of options and configure corresponding authentication parameters (more details and configuration options can be found from :ref:`lib_multicell_location`):

* :kconfig:`CONFIG_MULTICELL_LOCATION_SERVICE_NRF_CLOUD`

> * :kconfig:`CONFIG_MULTICELL_LOCATION_SERVICE_HERE` and :kconfig:`CONFIG_MULTICELL_LOCATION_HERE_API_KEY` (see below other authentication options)

Author: I need to check later if the whole text in parenthesis is just copy-paste from multicell and doesn't apply here. Or should refer to multicell documentation.

> * :kconfig:`CONFIG_MULTICELL_LOCATION_SERVICE_HERE` and :kconfig:`CONFIG_MULTICELL_LOCATION_HERE_API_KEY` (see below other authentication options)

Author: Removed the stuff from parenthesis.

      * The service is selected in the :c:struct:`location_method_config` structure when requesting for location.
      * The data transport method for the service is REST.
      * The services available are `nRF Cloud Location Services`_, `HERE Positioning`_, `Skyhook Precision Location`_ and `Polte Location API`_
* Wi-Fi positioning
   * Zephyr's Network Management API :ref:`zephyr:net_mgmt_interface` for getting the visible Wi-Fi access points.
   * Sending access point information to the selected location service and getting the calculated location back to the device:
      * The location library has an implementation for the Wi-Fi location services.
      * The service is selected in the :c:struct:`location_method_config` structure when requesting for location.
      * The data transport method for the service is REST.
      * The services available are `nRF Cloud Location Services`_, `HERE Positioning`_ and `Skyhook Precision Location`_.

Requirements
************

nRF Cloud certificates
======================

When using nRF Cloud for any location data, you must have the certificate provisioned.
See `Updating the nRF Cloud certificate`_ for more information.

> nRF9160 DK come pre-provisioned with certificates for nRF Cloud.

Reviewer 5: I just noticed a typo here. "come" should be "comes".

* :kconfig:`CONFIG_MULTICELL_LOCATION_SERVICE_SKYHOOK` and :kconfig:`CONFIG_MULTICELL_LOCATION_SKYHOOK_API_KEY`
* :kconfig:`CONFIG_MULTICELL_LOCATION_SERVICE_POLTE` and :kconfig:`CONFIG_MULTICELL_LOCATION_POLTE_CUSTOMER_ID` and :kconfig:`CONFIG_MULTICELL_LOCATION_POLTE_API_TOKEN`

For Wi-Fi location services, use at least one of the following sets of options and configure the corresponding authentication parameters:

* :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_NRF_CLOUD`
* :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_HERE` and :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_HERE_API_KEY`
* :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_SKYHOOK` and :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_SKYHOOK_API_KEY`

The following options are related to the Wi-Fi service and can usually have the default values:

* :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_HERE_HOSTNAME`
* :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_HERE_TLS_SEC_TAG`
* :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_SKYHOOK_HOSTNAME`
* :kconfig:`CONFIG_LOCATION_METHOD_WIFI_SERVICE_SKYHOOK_TLS_SEC_TAG`

Usage
*****

> To use the Location library, first initialize the library with the :c:func:`location_init` function.
> 
> Then, create the configuration (:c:struct:`location_config`) including the location method (:c:struct:`location_method_config`) and set the default values by passing them to :c:func:`location_config_defaults_set`.
> Location method configurations (:c:struct:`location_method_config`) should be passed to :c:func:`location_config_method_defaults_set` also.
> And then any required non-default values should be set to the structures.
> 
> Once you have set up the configuration properly, call the :c:func:`location_request` function with the configuration.

Reviewer 3: ```suggestion
To use the Location library, perform the following steps:

1. Initialize the library with the :c:func:`location_init` function.
#. Create the configuration (:c:struct:`location_config`) including the location method (:c:struct:`location_method_config`).
#. Set the default values by passing them to the :c:func:`location_config_defaults_set` function.
   Location method configurations (:c:struct:`location_method_config`) should be passed also to the :c:func:`location_config_method_defaults_set`  function.
#. Set any required non-default values to the structures.
#. Call the :c:func:`location_request` function with the configuration.
```
